internal/session/query.go

```diff
@@ -232,4 +234,49 @@ where
     )
 )
 `
+
+	// connectionsToCloseCte finds connections that are:
+	//
+	// * not closed
+	// * not announced by a given server in its latest update
+	//
+	// and marks them as closed.
+	connectionsToCloseCte = `
+with
+  -- Find connections that are not closed so we can reference those IDs
+  unclosed_connections as (
+    select connection_id
+      from session_connection_state
+    where
+      -- It's the current state
+      end_time is null
+        and
+      -- Current state isn't closed state
+      (state = 'authorized' or state = 'connected')
+        and
+      -- It's not in limbo between when it moved into this state and when
+      -- it started being reported by the worker, which is roughly every
+      -- 2-3 seconds
+      start_time < now() - interval '10 seconds'
+  ),
+  connections_to_close as (
+    select public_id
+      from session_connection
+    where
+      -- Related to the worker that just reported to us
+      server_id = $1
+        and
+      -- These are connection IDs that just got reported to us by the given
+      -- worker, so they should not be considered closed.
+        %s
+      -- Only unclosed ones
+      public_id in (select connection_id from unclosed_connections)
+  )
+  update session_connection
+    set
+      closed_reason = 'system error'
```

Comment:
"system error" seems to generic to me. Should we create a new `closed_reason` for this? Maybe "worker terminated"?

Reply:
I don't know that we can say for sure that the worker was terminated. We just know that the worker is no longer advertising it.

I think we should keep this as is, and I'll add another subtask to have the worker, when given a shutdown signal, attempt to gracefully terminate its connections and tell the controller that the reason is `worker terminating` or so.